### DISCUSSION TITLE
Use of keystore to protect local keys on Android

### DISCUSSION THREAD

**Discussion by scode** [standardnotes/forum#3199] [Suggestions]
**Have you read the [Longevity Statement](https://standardnotes.org/longevity)?**

Yes.

**Is your feature request related to a problem? Please describe.**

Today, it *appears* that after configuring StandardNotes on Android to require biometric unlock every time the app is accessed, it does not actually cryptographically protect the underlying data or remove it from memory.

As of a recent update that appears to have forgotten at least some settings (very noticeable on both desktop and Android as it went from dark mode into light mode), I was able to access my notes without going through biometric unlock. In other words:

1. Have standardnotes set to require biometric authentication immediately after switching away from the app.
2. Update occurred (presumably).
3. Started standardnotes and was presented with my data w/o biometric prompt.

I suspected this is because the biometric unlock feature is just a UI gate, rather than actually protecting the underlying data.

This was surprising to me given the nature of the product and its intended use case, but I found that this behavior [is in fact documented](https://standardnotes.com/help/79/how-does-standard-notes-encrypt-data-on-my-device).

**Describe the solution you'd like**

Ideally, the best available mechanism would be used to store a secret protected behind biometric authentication, leaving data encrypted on disk. From the settings menu, I believe this is what the passcode protection does (given its description).

For example precedence, see [the 1Password](https://support.1password.com/android-biometric-unlock-security/#your-1password-account-password-is-stored-securely) documentation. I am guessing they are using the [Android Keystore](https://developer.android.com/training/articles/keystore) which [appears to be providing support for this](https://developer.android.com/training/articles/keystore#UserAuthentication).

**Describe alternatives you've considered**

Clearly warn the user (right there in the settings page) that use of biometric locking does not prevent the data from being readable on the device (wording similar to the [afore mentioned page](https://standardnotes.com/help/79/how-does-standard-notes-encrypt-data-on-my-device) could be used). That way the user is aware of the behavior and can make an informed choice.

I do realize that the passcode setting explicitly says that it encrypts data. One could argue that the absence of such a statement next to the biometric lock setting is an indicator - but that's IMO too implicit.

**Comment by effieeee** [comment]
Hi, thank you for bringing this up! I believe one of the reasons why biometrics hasn't been made to encrypt local data yet is that the app already provides encryption for the keys and local data when the user creates an account for syncing notes, though I do understand this would be good to have in the case of offline users without accounts. Perhaps we can also add a clearer message on the UI to mention that the Biometrics lock cannot encrypt the local data 🙏

**Reply by scode** [reply]
Oh, the use case I'm concerned with is not account-less. It's just about whether the local physical attack vector requires that the secure enclave on the device be compromised, or whether all you need it something like a minor bug (such as the settings reset in this case), or compromising the file system while SN is not active. Basically, the ideal is that once the app is "locked", unencrypted data is no longer being stored in memory or or on disk, and the key required for decryption being in the secure enclave. If that were the case, there would, for example, be no way for a settings reset to reveal data since the data simply is not there (just like when you use the passcode option).

That said, I get that this takes resources to implement and maintain!
